src/main.rs

```diff
@@ -150,29 +239,58 @@ fn set(path: PathBuf, query: &str, value_str: &str) -> Result<(), Error> {
                 if n >= &len {
                     Err(CliError::ArrayIndexOob())?;
                 }
-                match &item { Item::Value(_) => already_inline = true, _ => () };
+
+                if let Item::Value(_) = &item {
+                    already_inline = true;
+                }
+
                 item = &mut item[n];
             }
             Name(n) => {
                 match &item {
                     Item::Table(_) => (),
                     Item::Value(Value::InlineTable(_)) => already_inline = true,
                     // TODO make this more directly construct the new, inner part?
-                    _ => *item = if already_inline {
-                        Item::Value(Value::InlineTable(Default::default()))
-                    } else {
-                        Item::Table(Table::new())
-                    },
+                    _ => {
+                        *item = if already_inline {
+                            Item::Value(Value::InlineTable(Default::default()))
+                        } else {
+                            Item::Table(Table::new())
+                        }
+                    }
                 };
                 item = &mut item[n];
             }
         }
     }
-    *item = value(value_str);
+    *item = detect_value(value_str);
 
-    // TODO actually write back
-    print!("{}", doc.to_string());
-    Ok(())
+    let result = if opts.overwrite {
+        // write content to path
+        if opts.backup {
+            let now: DateTime<Utc> = Utc::now();
+            let ext = now.format("%Y%m%d-%H%M%S-%f");
+            let backup_file = format!("{}.{}", path.display(), ext);
+            fs::copy(path.clone(), backup_file)?;
+        }
+        let mut output = OpenOptions::new().write(true).truncate(true).open(path)?;
+        write!(output, "{}", doc)?;
+        None
+    } else {
+        Some(format!("{}", doc))
+    };
+
+    Ok(result)
+}
+
+fn detect_value(value_str: &str) -> Item {
+    if let Ok(i) = value_str.parse::<i64>() {
+        value(i)
+    } else if let Ok(b) = value_str.parse::<bool>() {
+        value(b)
+    } else {
+        value(value_str)
```

Comment:
Hmm, this feels rather too magical to me for the interface of this command. In particular, this means that even when you're just trying to write string data, you'll sometimes have it come out as an int or a bool instead, just depending on the data -- depending on exactly what string it was you were trying to store.

Or for another angle on the same thing: this approach would mean that it'd be impossible to store some strings, like "1" or "false".

For example, it'd be impossible to correctly set `package.edition` in a `Cargo.toml` file:
```
$ target/debug/toml set --overwrite Cargo.toml package.edition 2018

$ git diff -U0
diff --git a/Cargo.toml b/Cargo.toml
index 2f232cf98..ac60e2325 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10 +10 @@ license = "MIT"
-edition = "2018"
+edition = 2018

$ cargo check
error: failed to parse manifest at `/home/greg/w/toml-cli/Cargo.toml`

Caused by:
  data did not match any variant of untagged enum MaybeWorkspace for key `package.edition`
```

Reply:
Instead, what would you think about having `toml set` take a flag to tell it how it should parse the value? Probably "string" would still be the default -- I feel like that's probably the most common.

So e.g.
```
toml set foo.toml x asdf   # a string
toml set foo.toml x --string asdf  # the same thing, explicitly

toml set foo.toml x 2018  # still a string: `x = "2018"`
toml set foo.toml x --int 2018  # an int: `x = 2018`
toml set foo.toml x --int asdf  # error

toml set foo.toml x --bool false
toml set foo.toml x --bool asdf  # error
```